pandas/core/groupby/generic.py

```diff
@@ -1259,7 +1259,14 @@ def value_counts(
         rep = partial(np.repeat, repeats=np.add.reduceat(inc, idx))
 
         # multi-index components
-        labels = list(map(rep, self.grouper.recons_labels)) + [llab(lab, inc)]
+        try:
+            labels = list(map(rep, self.grouper.recons_labels)) + [llab(lab, inc)]
+        except ValueError:
```

Comment:
can you give a self-contained example using *just* these routines showing how this fails; we *really* don't want to just catch it if it fails, rather we may need to adjust the logic in recons_labels.

Reply:
yes here is the mcve:
```python
import pandas as pd

s = pd.Series([1, 2], pd.DatetimeIndex(['2019-01', '2021-03']))
sg = s.groupby(pd.Grouper(freq='1Y'))
sg.value_counts()
```

and it goes fail like this:
```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-a41289f1d858> in <module>
      3 s = pd.Series([1, 2], pd.DatetimeIndex(['2019-01', '2021-03']))
      4 sg = s.groupby(pd.Grouper(freq='1Y'))
----> 5 sg.value_counts()

~/pandas-0xF4D3C0D3/pandas/core/groupby/generic.py in value_counts(self, normalize, sort, ascending, bins, dropna)
   1267 
   1268         # multi-index components
-> 1269         labels = list(map(rep, self.grouper.recons_labels)) + [llab(lab, inc)]
   1270         levels = [ping.group_index for ping in self.grouper.groupings] + [lev]
   1271         names = self.grouper.names + [self._selection_name]

~/.conda/envs/pandas-dev/lib/python3.7/site-packages/numpy/core/fromnumeric.py in repeat(a, repeats, axis)
    469 
    470     """
--> 471     return _wrapfunc(a, 'repeat', repeats, axis=axis)
    472 
    473 

~/.conda/envs/pandas-dev/lib/python3.7/site-packages/numpy/core/fromnumeric.py in _wrapfunc(obj, method, *args, **kwds)
     54 def _wrapfunc(obj, method, *args, **kwds):
     55     try:
---> 56         return getattr(obj, method)(*args, **kwds)
     57 
     58     # An AttributeError occurs if the object does not have

ValueError: operands could not be broadcast together with shape (3,) (2,)
```

because `rep` is like this: `functools.partial(<function repeat at 0x7f143c0eedd0>, repeats=array([1, 1]))` and `self.grouper.recons_labels` is like this: [array([0, 1, 2])]. Thus I tried making them match each other. But I found that the recons_labels is always well initialized except for groupby Grouper which has empty groups. I think if I can initialize recons_labels properly from the first, then it's good to do that way.

Reply:
oh wait I just found something. your comment gave a hint to me.

Reply:
I found that the length of `ids` and `self.recons_labels` is always the same except when the grouper has some periods which are empty periods. So I propose this:
```diff

```

it passes all tests and this issue. the behavior is the same as the current PR though, it's more specific than just ValueError.

Reply:
If I should correct the recons_labels logic, then at the logic, it should be able to know about empty periods. But recons_labels is under the grouper and grouper is already initialized without information about empty periods. So for addressing this issue, I tried making grouper be initialized without empty periods. then however this issue was solved but other tests went fail.

Reply:
The problem is that `ids` and `val` have only compact values, however, `recons_labels` have a all groups even including empty ones. So I think the workaround will be one of these: 1. make `ids` and `val` meet with `recons_labels` or 2. make `recons_labels` meet with `ids` and `val`. But this seems to matter only within `value_counts`. So rather than changing the logic in the recons_labels, I propose the second approach(make the result of `recons_labels` meet with `ids` and `val`)

Reply:
I think the post-mortem is a good moment to know whether the group has empty groups. But, If you want a more explicit moment, how about this?:
```diff
--- a/pandas/core/groupby/generic.py
+++ b/pandas/core/groupby/generic.py
@@ -54,6 +54,7 @@ from pandas.core.groupby.groupby import (
     _transform_template,
     groupby,
 )
+from pandas.core.groupby.ops import BinGrouper
 from pandas.core.index import Index, MultiIndex, _all_indexes_same
 import pandas.core.indexes.base as ibase
 from pandas.core.internals import BlockManager, make_block
@@ -1259,11 +1260,10 @@ class SeriesGroupBy(GroupBy):
         rep = partial(np.repeat, repeats=np.add.reduceat(inc, idx))

         # multi-index components
-        try:
-            labels = list(map(rep, self.grouper.recons_labels)) + [llab(lab, inc)]
-        except ValueError:
-            # If applying rep to recons_labels go fail, use unique ids
+        if isinstance(self.grouper, BinGrouper) and len(self.grouper.binlabels) != len(self.grouper.indices):
             labels = list(map(rep, [np.unique(ids)])) + [llab(lab, inc)]
+        else:
+            labels = list(map(rep, self.grouper.recons_labels)) + [llab(lab, inc)]
         levels = [ping.group_index for ping in self.grouper.groupings] + [lev]
         names = self.grouper.names + [self._selection_name]
```

Reply:
Hello @jreback, good morning how about my latest proposal? 

I also wanted to re-design the construction of recons_labels, but it has a lot of dependents and because of no comments, I couldn't do so yet.

Reply:
I don't want a try/except here at all.

Reply:
I got rid of try-except there and left only if statement.

Reply:
@jreback  What should I do next?